objects/src/accounts/vault.rs

```diff
@@ -0,0 +1,68 @@
+use super::{AccountError, AccountId, Asset, Digest};
+
+// ACCOUNT VAULT
+// ================================================================================================
+
+/// An asset container for an account.
+///
+/// An account vault can contain an unlimited number of assets. The assets are stored in a Sparse
+/// Merkle tree as follows:
+/// - For fungible assets, the index of a node is defined by the issuing faucet ID, and the value
+///   of the node is the asset itself. Thus, for any fungible asset there will be only one node
+///   in the tree.
+/// - For non-fungible assets, the index is defined by the asset itself, and the asset is also
+///   the value of the node.
+///
+/// An account vault can be reduced to a single hash which is the root of the Sparse Merkle tree.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct AccountVault {
+    // TODO: add backing sparse Merkle tree
+    assets: Vec<Asset>,
+}
+
+impl AccountVault {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new account vault initialized with the provided assets.
+    ///
+    /// TODO: return error if there are duplicates in the provided asset list.
+    pub fn new(assets: &[Asset]) -> Self {
```

Comment:
i think `new` constructors should always reflect what they store internally (in this case, `Vec<Asset>`) so it's up to the user how he is going to build it.

if we ask for a slice, it might be that the user will have to allocate an array of its own, just for it to be reallocated inside of the function.

if the goal is to increase the flexibility for the user, we might as well introduce a generic here that will build a `Vec` (i.e. `IntoIterator`)

Reply:
As I mentioned in a coupe of other comments - I'm not yet sure what's the best way to handle both `AccountVault` and `AccountStorage`. The main issue is that these could potentially get really big (i.e., a vault with millions of assets) and maybe in such cases they should be backed by a database rather than a simple in-memory structure.

So, this `new()` method should be viewed more as illustrative to show that we can construct a vault from a list of assets and I didn't want to overcomplicate it with the iterator just yet.